plugins/rtmp-services/service-specific/twitch.c

```diff
@@ -7,31 +7,53 @@
 
 #include "twitch.h"
 
-static update_info_t *twitch_update_info = NULL;
-static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
-static bool ingests_refreshed = false;
-static bool ingests_refreshing = false;
-static bool ingests_loaded = false;
-
 struct ingest {
 	char *name;
 	char *url;
+	char *rtmps_url;
+};
+
+struct service_ingests {
+	update_info_t *update_info;
+	pthread_mutex_t mutex;
+	bool ingests_refreshed;
+	bool ingests_refreshing;
+	bool ingests_loaded;
+	DARRAY(struct ingest) cur_ingests;
+	const char *cache_old_filename;
+	const char *cache_new_filename;
 };
 
-static DARRAY(struct ingest) cur_ingests;
+#define INGESTS_INITIALIZER(cache_old_filename_, cache_new_filename_)    \
+	{                                                                \
+		.update_info = NULL, .mutex = PTHREAD_MUTEX_INITIALIZER, \
+		.ingests_refreshed = false, .ingests_refreshing = false, \
+		.ingests_loaded = false, .cur_ingests = {0},             \
+		.cache_old_filename = cache_old_filename_,               \
+		.cache_new_filename = cache_new_filename_                \
+	}
```

Comment:
Yeah I really don't see a benefit of this over directly declaring the static structs.

If the idea is to create a generic API to create IVS ingest structs, it should be a function that returns a generated struct, but I don't think it's generalised this much.

Reply:
I'd prefer not to duplicate this code, or keep everything except for the cache filenames deduplicated somehow; not entirely sure about `PTHREAD_MUTEX_INITIALIZER` but I guess all this initialization could also happen in a normal function instead (though it would happen at a different time compared to this)

Reply:
> Well the code _is_ duplicated, because that's what the preprocessor will do and there's no benefit to hiding that from developers (the only benefit was in effort of _writing_ the source code, but in general code should be optimised for the reader, not the writer).

in this case I'd see it as benefit to the reader (and future maintainer) to not have to read the (mostly) same struct definition twice (or modify it twice, for future maintainers); unfortunately this is a macro and not e.g. a c++ constexpr (or similar) function, so deduplication is harder to argue if people look at the generated source code I guess?

Reply:
changed these to expanded versions in https://github.com/obsproject/obs-studio/pull/10634#commits-pushed-9681a0a; not sure if the fixup commits apply cleanly, though fixing them up should be relatively straightforward

Reply:
Manually cleaned these up.